react-big-calendar/index.d.ts

```diff
@@ -26,10 +26,10 @@ declare module 'react-big-calendar' {
         step?: number;
         rtl?: boolean;
         eventPropGetter?: (event: Object, start: stringOrDate, end: stringOrDate, isSelected: boolean) => void;
-        titleAccessor?: string;
-        allDayAccessor?: boolean;
-        startAccessor?: stringOrDate;
-        endAccessor?: stringOrDate;
+        titleAccessor?: string | ((row: Object) => string);
+        allDayAccessor?: string | ((row: Object) => boolean);
+        startAccessor?: string | ((row: Object) => Date);
+        endAccessor?: string | ((row: Object) => Date);
```

Comment:
Okay, there are now two issues present in the type guard you've provided.  I agree that the output of the function should be a Date.  

Firstly, the first portion will still need to be a string, because all property names of objects are stored as strings.  There's a quick paragraph called "Property names" in the following link:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Property_Accessors

Secondly, now you've introduced a spread operator into your function signature, which is not necessary.  Below are a couple of links to the usage of startAccessor in the calendar:
https://github.com/intljusticemission/react-big-calendar/blob/fa73913d8f1198359aa06256f4e5dd88557d8ef1/src/utils/eventLevels.js#L13
https://github.com/intljusticemission/react-big-calendar/blob/fa73913d8f1198359aa06256f4e5dd88557d8ef1/src/utils/accessors.js#L6

Looking at the first link, pay particular attention to the **dates.startOf(get(event, startAccessor), 'day'**, you'll see "get" which is an alias for "assessor" in the import statement on line 2.  Note that the first argument being passed into "get" is an event object and the second argument is the startAccessor that we're having a conversation about.

Now have a look at the second link, the acessor.js file.  Note that on line 6, "field" is our startAccessor function and it's being handed a parameter of "data" which is our event object.  There is only one parameter, which is the event object.  In other words, there's no need for a spread operator of any's (...any[]).  In fact, we could go even further with our type inference and avoid using "any" altogether.